log/loki.go

```diff
@@ -0,0 +1,499 @@
+	pushPeriod time.Duration
+	client     *http.Client
+	ctx        context.Context
+	profile    bool
+}
+
+// LokiFromConfigLine returns a new logrus.Hook that pushes logrus.Entrys to loki and is configured
+// through the provided line
+func LokiFromConfigLine(ctx context.Context, line string) (logrus.Hook, error) {
+	h := &lokiHook{
+		addr:       "http://127.0.0.1:3100/loki/api/v1/push",
+		limit:      100,
+		levels:     logrus.AllLevels,
+		pushPeriod: time.Second * 1,
+		ctx:        ctx,
+		msgMaxSize: 1024 * 1024, // 1mb
+	}
+	if line == "loki" {
+		return h, nil
+	}
+
+	parts := strings.SplitN(line, "=", 2)
+	if parts[0] != "loki" {
+		return nil, fmt.Errorf("loki configuration should be in the form `loki=url-to-push` but is `%s`", line)
+	}
+	args := strings.Split(parts[1], ",")
+	h.addr = args[0]
+	// TODO use something better ... maybe
+	// https://godoc.org/github.com/kubernetes/helm/pkg/strvals
+	// atleast until https://github.com/loadimpact/k6/issues/926?
+	if len(args) == 1 {
+		return h, nil
+	}
+
+	for _, arg := range args[1:] {
+		paramParts := strings.SplitN(arg, "=", 2)
+
+		if len(paramParts) != 2 {
+			return nil, fmt.Errorf("loki arguments should be in the form `address,key1=value1,key2=value2`, got %s", arg)
+		}
+
+		key := paramParts[0]
+		value := paramParts[1]
+
+		var err error
+		switch key {
+		case "pushPeriod":
+			h.pushPeriod, err = time.ParseDuration(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki pushPeriod %w", err)
+			}
+		case "profile":
+			h.profile = true
+		case "limit":
+			h.limit, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki limit as a number %w", err)
+			}
+		case "msgMaxSize":
+			h.msgMaxSize, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki msgMaxSize as a number %w", err)
+			}
+		case "level":
+			h.levels, err = getLevels(value)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			if strings.HasPrefix(key, "label.") {
+				labelKey := strings.TrimPrefix(key, "label.")
+				h.labels = append(h.labels, [2]string{labelKey, value})
+				continue
+			}
+
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.client = &http.Client{Timeout: h.pushPeriod}
+
+	h.start()
+	return h, nil
+}
+
+func getLevels(level string) ([]logrus.Level, error) {
+	// TODO figure out if `tracing`,`fatal` and `panic` should be included
+	levels := []logrus.Level{}
+	switch level {
+	case "debug":
+		levels = append(levels, logrus.DebugLevel)
+		fallthrough
+	case "info":
+		levels = append(levels, logrus.InfoLevel)
+		fallthrough
+	case "warning":
+		levels = append(levels, logrus.WarnLevel)
+		fallthrough
+	case "error":
+		levels = append(levels, logrus.ErrorLevel)
+	default:
+		return nil, fmt.Errorf("unknown log level %s", level)
+	}
+	return levels, nil
+}
+
+func (h *lokiHook) start() {
+	h.ch = make(chan *logrus.Entry, 1000)
+	go h.loop()
```

Comment:
Minor nitpick: this whole `start()` method is superfluous - you can `make()` the channel in the constructor and then directly call `go h.loop()` instead of `h.start()`